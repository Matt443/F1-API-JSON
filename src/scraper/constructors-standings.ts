import axios from "axios"
import cheerio from "cheerio"

import { dynamicLinks } from "../endpoints/endpoints"

import {constructorStanding} from "../types/types"

export const getConstructorStandings = (year: number = new Date().getFullYear()): Promise<constructorStanding[]> => {

    let constructorStandings: constructorStanding[] = []

    return new Promise((resolve, reject) => {
        axios(`${dynamicLinks.constructorStandings_1}/${year}/${dynamicLinks.constructorStandings_2}`)
            .then(
                response => {
                    const html = response.data
                    const $ = cheerio.load(html)

                    $('tr').each(
                        function () {
                            const position: number = parseInt($(this).find('td:nth-child(2)').text().trim())
                            const team: string = $(this).find('td:nth-child(3)').text().trim()
                            const points: number = parseInt($(this).find('td:nth-child(4)').text().trim())

                            if (!Number.isNaN(position) && team.length !== 0 && !Number.isNaN(points)) {
                                const constructorStanding: constructorStanding = {
                                    position,
                                    team,
                                    points
                                }
                                constructorStandings.push(constructorStanding)
                            }
                        })
                    resolve(constructorStandings)
                })
            .catch(err => { reject(err) })
    })

}