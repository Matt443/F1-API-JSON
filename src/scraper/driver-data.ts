import axios from "axios";
import cheerio from "cheerio";

import { staticLinks } from "../endpoints/endpoints";

import { drivers } from "../types/types";

export const getDriverData = (): Promise<drivers[]> => {
    try {
        let drivers: drivers[] = [];

        return new Promise(async (resolve, reject) => {
            const response = await axios(staticLinks.drivers);
            const $ = cheerio.load(response.data);

            $("fieldset").each(function () {
                const firstName: string = $(this).find(".container > div:nth-child(1) > div:nth-child(1) > span:nth-child(1)").text().trim();
                const secondName: string = $(this).find(".container > div:nth-child(1) > div:nth-child(1) > span:nth-child(2)").text().trim();
                const team: string = $(this).find(" p:nth-child(3)").text();
                const rank: number = parseInt($(this).find(".rank").text());
                const points: number = parseInt($(this).find(".points > div:nth-child(1)").text());
                const nationalityImage: string | undefined = $(this)
                    .find(".container > .row.justify-content-between.align-items-center.listing-item--head > .col-xs-4.country-flag > picture:nth-child(1) > img:nth-child(2)")
                    .attr("data-src");
                const driverImage: string | undefined = $(this).find(" div:nth-child(4) > picture:nth-child(1) > img:nth-child(2)").attr("data-src");

                if (firstName.length !== 0 && secondName.length !== 0 && team.length !== 0 && !Number.isNaN(rank) && !Number.isNaN(points)) {
                    const driver: drivers = {
                        name: firstName.concat(" ", secondName),
                        team,
                        rank,
                        points,
                        nationalityImage,
                        driverImage,
                    };
                    drivers.push(driver);
                }
            });
            resolve(drivers);
        });
    } catch (error: any) {
        throw new Error(error);
    }
};
