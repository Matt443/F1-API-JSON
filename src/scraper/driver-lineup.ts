import axios from "axios";
import cheerio from "cheerio";

import { staticLinks } from "../endpoints/endpoints";

import { isDriver } from "../types/types";

export const getDriverLineup = async (): Promise<isDriver[]> => {
    try {
        let drivers: isDriver[] = [];

        const response = await axios(staticLinks.drivers);
        const $ = cheerio.load(response.data);

        $("a.outline").each(function () {
            const firstName: string = $(this).find("div:nth-child(1) > div:nth-child(1) > div:nth-child(3) > h4:nth-child(1) > div:nth-child(1) > p:nth-child(1)").text().trim();
            const secondName: string = $(this).find("div:nth-child(1) > div:nth-child(1) > div:nth-child(3) > h4:nth-child(1) > div:nth-child(1) > p:nth-child(2)").text().trim();
            const team: string = $(this).find("div:nth-child(1) > div:nth-child(1) > p:nth-child(5)").text();
            const rank: string | undefined = $(this).find("div:nth-child(1) > div:nth-child(1) > div:nth-child(1) > p:nth-child(1)").text().trim();
            const nationalityImage: string | undefined = $(this).find("div:nth-child(1) > div:nth-child(1) > div:nth-child(3) > img:nth-child(2)").attr("src");
            const driverImage: string | undefined = $(this).find("div:nth-child(1) > div:nth-child(1) > div:nth-child(6) > img:nth-child(2)").attr("src");

            if (firstName.length !== 0 && secondName.length !== 0 && team.length !== 0) {
                const driver: isDriver = {
                    name: firstName.concat(" ", secondName),
                    team,
                    rank,
                    nationalityImage,
                    driverImage,
                };
                drivers.push(driver);
            }
        });
        if (drivers.length === 0) {
            throw new Error("No data found");
        }
        return drivers;
    } catch (error: any) {
        throw new Error(error);
    }
};
