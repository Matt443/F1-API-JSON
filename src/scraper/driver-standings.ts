import axios from "axios"
import cheerio from "cheerio"

import { dynamicLinks } from "../endpoints/endpoints"

interface driverStanding {
    position: number
    driver: string
    nationality: string
    team: string
    points: number
}

export const getDriverStandings = (year: number): Promise<driverStanding[]> => {

    let driverStandings: driverStanding[] = []

    return new Promise((resolve, reject) => {
        axios(`${dynamicLinks.driverStandings_1}/${year}/${dynamicLinks.driverStandings_2}`)
            .then(
                response => {
                    const html = response.data
                    const $ = cheerio.load(html)

                    $('tr').each(
                        function () {
                            const position: number = parseInt($(this).find(' td:nth-child(2) ').text())

                            const firstName: string = $(this).find(' a.dark.bold.ArchiveLink > span:nth-child(1)').text()
                            const lastName: string = $(this).find(' a.dark.bold.ArchiveLink > span:nth-child(2)').text()

                            const driver = firstName.concat(" ", lastName)
                            const nationality: string = $(this).find(' td.dark.semi-bold.uppercase').text()
                            const team: string = $(this).find('td:nth-child(5) > a.grey.semi-bold.uppercase.ArchiveLink').text()
                            const points: number = parseInt($(this).find(' td:nth-child(6) ').text())

                            if (position !== NaN && points !== NaN && driver.length !== 0 && nationality.length !== 0 && team.length !== 0) {
                                const driverStanding: driverStanding = {
                                    position,
                                    driver,
                                    nationality,
                                    team,
                                    points
                                }
                                driverStandings.push(driverStanding)
                            }
                        })
                    resolve(driverStandings)
                    return Promise.all(driverStandings)
                })
            .catch(err => { reject(err) })
    })
}