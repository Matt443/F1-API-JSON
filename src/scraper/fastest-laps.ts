import axios from "axios";
import * as cheerio from "cheerio";

import { dynamicLinks } from "../endpoints/endpoints";

import { isFastestLap } from "../types/types";

export const getFastestLaps = async (year: number = new Date().getFullYear()): Promise<isFastestLap[]> => {
    try {
        let fastestLaps: isFastestLap[] = [];

        const response = await axios(`${dynamicLinks.rootLink}/${year}/${dynamicLinks.fastestLap}`);
        const $ = cheerio.load(response.data);

        $(".f1-table > tbody:nth-child(2) > tr").each(function () {
            const grandPrixAll: string = $(this).find("td:nth-child(1) > p:nth-child(1) > a").contents().get()[1];
            const grandPrix: string = $(grandPrixAll).text();
            const driverName: string = $(this)
                .find("td:nth-child(2) > p:nth-child(1) > span:nth-child(1)")
                .text()
                .trim()
                .replace(/\u00a0/g, " ");
            const car: string = $(this).find("td:nth-child(3) > p:nth-child(1)").text().trim();
            const time: string = $(this).find("td:nth-child(4) > p:nth-child(1)").text().trim();

            const driver: string = driverName.slice(0, driverName.length - 3);

            if (driver.length !== 0 && grandPrix.length !== 0 && car.length !== 0 && time.length !== 0) {
                const fastestLap: isFastestLap = {
                    driver,
                    grandPrix,
                    car,
                    time,
                };

                fastestLaps.push(fastestLap);
            }
        });

        if (fastestLaps.length === 0) {
            throw new Error("No data found");
        }

        return fastestLaps;
    } catch (error: any) {
        throw new Error(error);
    }
};
