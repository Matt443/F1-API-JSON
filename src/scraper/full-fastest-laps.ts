import { staticLinks } from "../endpoints/endpoints";
import { getResultURL } from "../utils/scrapping";

export async function getFullFastestLaps(year: number = new Date().getFullYear(), raceName: string = "Australia") {
    const resultsURL = await getResultURL(year, raceName);

    const fastestLapsURL = `${staticLinks.fullResults}/${year}/${resultsURL.slice(23, resultsURL.length - 11)}qualifying`;
    console.log();
}
