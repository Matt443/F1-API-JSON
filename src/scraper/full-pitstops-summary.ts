import axios from "axios";
import * as cheerio from "cheerio";
import { staticLinks } from "../endpoints/endpoints";

import { isPitstopsSummary } from "../types/types";
import { getF1Table, getResultURL } from "../utils/scrapping";

/**
 *
 * @param {year} year - 1950-now
 * @param {string} [raceName="Australia"]
 * @returns {Promise<isPitstopsSummary[]>}
 */

export const getFullPitstopsSummary = async (year: number = new Date().getFullYear(), raceName: string = "Australia"): Promise<isPitstopsSummary[]> => {
    try {
        const resultsURL = await getResultURL(year, raceName);
        const pitstopsSummaryURL = `${staticLinks.fullResults}/${year}/${resultsURL.slice(23, resultsURL.length - 12)}/pit-stop-summary`;

        const results = await axios(pitstopsSummaryURL);
        const $ = cheerio.load(results.data);
        const raceDate = $(".f1rd-page > div:nth-child(3) > div:nth-child(2) > div:nth-child(1) p:nth-child(1)").text().slice(5);

        function assignTableValues(driver: string[]): isPitstopsSummary {
            const driverObj: isPitstopsSummary = {
                stops: driver[0],
                name: driver[2].slice(0, driver[2].length - 3).replace(/\u00a0/g, " "),
                code: driver[2].slice(driver[2].length - 3),
                team: driver[3],
                lap: Number(driver[4]),
                time: driver[6],
                timeOfDay: new Date(`${raceDate} ${driver[5]} GMT-0`),
                number: Number(driver[1]),
                total: driver[7],
            };
            return driverObj;
        }

        return getF1Table(pitstopsSummaryURL, assignTableValues) as unknown as isPitstopsSummary[];
    } catch (error: unknown) {
        throw new Error(error as string);
    }
};
