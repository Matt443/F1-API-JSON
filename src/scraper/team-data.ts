import axios from "axios";
import cheerio from "cheerio"

import { links } from "../links/links"

interface teams {
    name: string
    drivers: string[]
    points: number
    rank: number
    carLogo: string | undefined
    carImage: string | undefined
}

export function getTeamsData() {

    let teamsArray: teams[] = []

    axios(links.teams).then(response => {
        const html = response.data
        const $ = cheerio.load(html)

        $('fieldset').each(
            function () {
                const name: string = $(this).find(' div:nth-child(1) > .listing-info > .name.f1-bold--m > .f1-color--black').text()
                const points: number = parseInt($(this).find('.points > div:nth-child(1)').text())
                const rank: number = parseInt($(this).find('.rank').text())

                const driver1_0: string = $(this).find(' div:nth-child(1) > div:nth-child(3) > div:nth-child(1) > div:nth-child(1) > span:nth-child(1)').text()
                const driver1_1: string = $(this).find(' div:nth-child(1) > div:nth-child(3) > div:nth-child(1) > div:nth-child(1) > span:nth-child(2)').text()
                const driver2_0: string = $(this).find('div:nth-child(1) > div:nth-child(3) > div:nth-child(2) > div:nth-child(1) > span:nth-child(1)').text()
                const driver2_1: string = $(this).find('div:nth-child(1) > div:nth-child(3) > div:nth-child(2) > div:nth-child(1) > span:nth-child(2)').text()
                const driver1 = driver1_0.concat(" ", driver1_1)
                const driver2 = driver2_0.concat(" ", driver2_1)
                const drivers: string[] = [driver1, driver2]

                const carLogo: string | undefined = $(this).find('div:nth-child(1) > div:nth-child(2) > div:nth-child(2) > picture:nth-child(1) > img:nth-child(2)').attr('data-src')
                const carImage: string | undefined = $(this).find('div:nth-child(1) > div:nth-child(4) > picture:nth-child(1) > img:nth-child(5)').attr('data-src')

                if (name.length !== 0 && points !== NaN && rank !== NaN) {
                    const teamObj: teams = {
                        name,
                        drivers,
                        points,
                        rank,
                        carLogo,
                        carImage,
                    }
                    teamsArray.push(teamObj)
                }
            })
        console.log(teamsArray)
    })
}