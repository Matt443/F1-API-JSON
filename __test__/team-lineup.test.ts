import { getTeamsData } from "../src/scraper/team-data"

test('current teams data' , async () => {
    expect(
        await getTeamsData()
    ).toMatchSnapshot()
})