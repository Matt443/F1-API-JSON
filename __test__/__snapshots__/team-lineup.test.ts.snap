// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`current teams data 1`] = `
[
  {
    "carImage": "https://www.formula1.com/content/dam/fom-website/teams/2022/red-bull-racing.png.transform/4col/image.png",
    "carLogo": "https://www.formula1.com/content/dam/fom-website/teams/2022/red-bull-racing-logo.png.transform/2col/image.png",
    "drivers": [
      "Sergio Perez",
      "Max Verstappen",
    ],
    "name": "Red Bull Racing",
    "points": 619,
    "rank": 1,
  },
  {
    "carImage": "https://www.formula1.com/content/dam/fom-website/teams/2022/ferrari.png.transform/4col/image.png",
    "carLogo": "https://www.formula1.com/content/dam/fom-website/teams/2022/ferrari-logo.png.transform/2col/image.png",
    "drivers": [
      "Carlos Sainz",
      "Charles Leclerc",
    ],
    "name": "Ferrari",
    "points": 454,
    "rank": 2,
  },
  {
    "carImage": "https://www.formula1.com/content/dam/fom-website/teams/2022/mercedes.png.transform/4col/image.png",
    "carLogo": "https://www.formula1.com/content/dam/fom-website/teams/2022/mercedes-logo.png.transform/2col/image.png",
    "drivers": [
      "George Russell",
      "Lewis Hamilton",
    ],
    "name": "Mercedes",
    "points": 387,
    "rank": 3,
  },
  {
    "carImage": "https://www.formula1.com/content/dam/fom-website/teams/2022/alpine.png.transform/4col/image.png",
    "carLogo": "https://www.formula1.com/content/dam/fom-website/teams/2022/alpine-logo.png.transform/2col/image.png",
    "drivers": [
      "Fernando Alonso",
      "Esteban Ocon",
    ],
    "name": "Alpine",
    "points": 143,
    "rank": 4,
  },
  {
    "carImage": "https://www.formula1.com/content/dam/fom-website/teams/2022/mclaren.png.transform/4col/image.png",
    "carLogo": "https://www.formula1.com/content/dam/fom-website/teams/2022/mclaren-logo.png.transform/2col/image.png",
    "drivers": [
      "Lando Norris",
      "Daniel Ricciardo",
    ],
    "name": "McLaren",
    "points": 130,
    "rank": 5,
  },
  {
    "carImage": "https://www.formula1.com/content/dam/fom-website/teams/2022/alfa-romeo.png.transform/4col/image.png",
    "carLogo": "https://www.formula1.com/content/dam/fom-website/teams/2022/alfa-romeo-logo.png.transform/2col/image.png",
    "drivers": [
      "Valtteri Bottas",
      "Guanyu Zhou",
    ],
    "name": "Alfa Romeo",
    "points": 52,
    "rank": 6,
  },
  {
    "carImage": "https://www.formula1.com/content/dam/fom-website/teams/2022/aston-martin.png.transform/4col/image.png",
    "carLogo": "https://www.formula1.com/content/dam/fom-website/teams/2022/aston-martin-logo.png.transform/2col/image.png",
    "drivers": [
      "Lance Stroll",
      "Sebastian Vettel",
    ],
    "name": "Aston Martin",
    "points": 45,
    "rank": 7,
  },
  {
    "carImage": "https://www.formula1.com/content/dam/fom-website/teams/2022/haas-f1-team.png.transform/4col/image.png",
    "carLogo": "https://www.formula1.com/content/dam/fom-website/teams/2022/haas-f1-team-logo.png.transform/2col/image.png",
    "drivers": [
      "Kevin Magnussen",
      "Mick Schumacher",
    ],
    "name": "Haas F1 Team",
    "points": 34,
    "rank": 8,
  },
  {
    "carImage": "https://www.formula1.com/content/dam/fom-website/teams/2022/alphatauri.png.transform/4col/image.png",
    "carLogo": "https://www.formula1.com/content/dam/fom-website/teams/2022/alphatauri-logo.png.transform/2col/image.png",
    "drivers": [
      "Pierre Gasly",
      "Yuki Tsunoda",
    ],
    "name": "AlphaTauri",
    "points": 34,
    "rank": 9,
  },
  {
    "carImage": "https://www.formula1.com/content/dam/fom-website/teams/2022/williams.png.transform/4col/image.png",
    "carLogo": "https://www.formula1.com/content/dam/fom-website/teams/2022/williams-logo.png.transform/2col/image.png",
    "drivers": [
      "Nicholas Latifi",
      "Alexander Albon",
    ],
    "name": "Williams",
    "points": 8,
    "rank": 10,
  },
]
`;
